bundles/org.openhab.binding.roborock/src/main/java/org/openhab/binding/roborock/internal/RoborockVacuumHandler.java

```diff
@@ -0,0 +1,1015 @@
+                            .isJsonObject()) {
+                JsonArray routinesArray = JsonParser.parseString(routinesResponse).getAsJsonObject().get("result")
+                        .getAsJsonArray();
+                Map<String, Object> routines = new HashMap<>();
+                for (int i = 0; i < routinesArray.size(); ++i) {
+                    JsonObject routinesJsonObject = routinesArray.get(i).getAsJsonObject();
+                    routines.put(routinesJsonObject.get("id").getAsString(),
+                            routinesJsonObject.get("name").getAsString());
+                }
+                updateState(CHANNEL_ROUTINES, new StringType(gson.toJson(routines)));
+            } else {
+                logger.debug("Routines not supported for device {}", config.duid);
+                supportsRoutines = false;
+            }
+        }
+
+        lastSuccessfulPollTimestamp = System.currentTimeMillis();
+        scheduleNextPoll(-1);
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    private void sendAllMqttCommands() {
+        try {
+            outstandingRequests.put("getStatus", sendRPCCommand(COMMAND_GET_STATUS));
+            outstandingRequests.put("getConsumable", sendRPCCommand(COMMAND_GET_CONSUMABLE));
+            outstandingRequests.put("getNetworkInfo", sendRPCCommand(COMMAND_GET_NETWORK_INFO));
+            outstandingRequests.put("getCleanSummary", sendRPCCommand(COMMAND_GET_CLEAN_SUMMARY));
+            outstandingRequests.put("getDndTimer", sendRPCCommand(COMMAND_GET_DND_TIMER));
+            outstandingRequests.put("getRoomMapping", sendRPCCommand(COMMAND_GET_ROOM_MAPPING));
+            outstandingRequests.put("getSegmentStatus", sendRPCCommand(COMMAND_GET_SEGMENT_STATUS));
+            outstandingRequests.put("getMapStatus", sendRPCCommand(COMMAND_GET_MAP_STATUS));
+            outstandingRequests.put("getLedStatus", sendRPCCommand(COMMAND_GET_LED_STATUS));
+            outstandingRequests.put("getCarpetMode", sendRPCCommand(COMMAND_GET_CARPET_MODE));
+            outstandingRequests.put("getFwFeatures", sendRPCCommand(COMMAND_GET_FW_FEATURES));
+            outstandingRequests.put("getMultiMapsList", sendRPCCommand(COMMAND_GET_MULTI_MAP_LIST));
+            outstandingRequests.put("getCustomizeCleanMode", sendRPCCommand(COMMAND_GET_CUSTOMIZE_CLEAN_MODE));
+        } catch (UnsupportedEncodingException e) {
+            logger.warn("Failed to send MQTT commands due to unsupported encoding: {}", e.getMessage());
+        }
+    }
+
+    public void handleMessage(byte[] payload) {
+        logger.debug("Received MQTT message for: {}", config.duid);
+        try {
+            String response = ProtocolUtils.handleMessage(payload, localKey, nonce);
+            if (response.isEmpty()) {
+                logger.debug("MQTT message processed - invalid message format received");
+                return;
+            }
+
+            logger.trace("Received MQTT message: {}", response);
```

Comment:
Sort of expensive. Can be surrounded with something like logger.isTraceEnabled()

Reply:
Done.